### PR TITLE
Fix parametric datatype instantiation

### PR DESCRIPTION
**Pull request by ajreynol**
A bug was introduced when adding the Node-level datatype implementation in https://github.com/CVC4/CVC4/commit/d803e7fcf60f9bb847853fe6ccf7589b94b76922. The code did not probably get the arity of a sort constructor.  This adds TypeNode::getSortConstructorArity and uses it during parametric datatype type resolution.

### CODE REVIEW

File: src/expr/type_node.cpp [hunk 1]
```diff
@@ -459,9 +459,16 @@ TypeNode TypeNode::instantiateParametricDatatype(
   return nm->mkTypeNode(kind::PARAMETRIC_DATATYPE, paramsNodes);
 }
 
+unsigned TypeNode::getSortConstructorArity() const
```
Comment (4tXJ7f) [inline src/expr/type_node.cpp:462]:
```suggestion
uint64_t TypeNode::getSortConstructorArity() const
```

The `SortArityAttr` atrribute maps to `uint64_t`. We could also consider returning a `size_t` here, I have no strong opinions.

File: src/expr/type_node.h [hunk 2]
```diff
@@ -674,6 +674,9 @@ class TypeNode {
   /** Is this a sort constructor kind */
   bool isSortConstructor() const;
 
+  /** Get sort constructor arity */
+  unsigned getSortConstructorArity() const;
```
Comment (4tXJ7f) [inline src/expr/type_node.h:678]:
```suggestion
  uint64_t getSortConstructorArity() const;
```

File: src/expr/dtype_cons.cpp [hunk 3]
```diff
@@ -595,9 +608,9 @@ TypeNode DTypeConstructor::doParametricSubstitution(
     children.push_back(
         doParametricSubstitution((*i), paramTypes, paramReplacements));
   }
-  for (size_t i = 0, psize = paramTypes.size(); i < psize; ++i)
+  for (unsigned i = 0; i < paramTypes.size(); ++i)
```
Comment (4tXJ7f) [inline src/expr/dtype_cons.cpp:611]: Why this change?

Comment (ajreynol) [reply src/expr/dtype_cons.cpp:611]: This was unintentional, ill revert.
